### DISCUSSION TITLE
Typescript: Type of Action Parameter in Union

### DISCUSSION THREAD

**Discussion by chriszwickerocteris** [mobxjs/mobx-state-tree#2150] [Q&A] [answered]
**_Bug report_**

* [x] I've checked documentation and searched for existing issues
* [x] I've made sure my project is based on the latest MST version (checked with `3.14.1`)
* [x] <del>Fork [this](https://codesandbox.io/s/y64pzxj01) code sandbox or</del> another minimal reproduction. 

Also, I've asked this as a [question](https://stackoverflow.com/q/57509041/7159043) on stackoverflow.

**Sandbox link or minimal reproduction code**

```typescript
const Foo = types.model({
  id: types.identifier,
  payload: types.maybe(types.string),
}).actions((self) => ({
  updatePayload(payload?: string) {
    self.payload = payload;
  },
}));

const Bar = types.model({
  id: types.identifier,
  payload: types.array(types.array(types.boolean)),
}).actions((self) => ({
  updatePayload(payload: boolean[][]) {
    applySnapshot(self.payload, payload);
  },
}));

const Item = types.union(Foo, Bar);
const Bucket = types.model({
  items: types.map(Item),
}).actions((self) => ({
  updateItemPayload(id: string, payload?: string | boolean[][]) {
    const item = self.items.get(id);
    item && item.updatePayload(payload);
  },
}));
```

**Describe the expected behavior**
`Item#updatePayload` should (I believe) expect a parameter of type `I extends SnapshotOrInstance<typeof Foo> ? string | undefined : boolean[][]`

**Describe the observed behavior**
`Item#updatePayload` expects a parameter of type `string & boolean[][]`, which of course is impossible.

**Comment by coolsoftwaretyler** [comment]
Hey @chriszwickerocteris - I'm sorry no one ever got back to you here. We are trying to revamp the MST repository, and TypeScript issues are definitely on our radar. 

I'm going to tag this as such. If this issue is still relevant to you and you're interested in helping, let me know. But I totally understand if it's just not a priority for ya anymore.

Thanks!

**Comment by chriszwickerocteris** [comment]
Hi @coolsoftwaretyler - we finally moved away from MST, so at least for now, this isn't an issue for me any longer. Thanks anyway ;-)

**Comment by coolsoftwaretyler** [comment]
Thanks for the follow up @chriszwickerocteris, sorry it didn't work out for ya.

We'll leave the issue open as we work to revamp everything

**Comment by thegedge** [comment, marked as answer]
Since I've been picking up some of the TS work, I can pitch in.

Both MST and TS are doing the right thing. In `updateItemPayload`, `item.updatePayload` is going to be of this type:

```ts
type F = ((payload: boolean[][]) => void) | ((payload: string) => void)
```

It doesn't know which of those two the function is, so what happens if we give `F` a `string` and the underlying instance is actually the function that takes a `boolean[][]`? That's why you're seeing `boolean[][] & string`, because if the payload you're providing is both possible types at the same time, it'll work for all members of the union. A simplified version of this in the [TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeUCuYAmBDYEAKWIANgPZYYAUYhp5AXFAEYklERYB2A2gLq8CUDAG4kAlhgDcUAL4AoUJCgAhBMjSYc+GmUrViOhgGdgAJ1EcA5oKgjxUuQugAxVRTgAfZfy4AidNlwCfXIfHglZWQBjEg5jKAAzBhdEDlQiIigsQ0yOEHCI+IofXGMffiA):

![typescript playground screenshot showing a similar error](https://github.com/mobxjs/mobx-state-tree/assets/833576/63ec1244-f709-4a28-8394-c1352bc35c1c)

Given your example, you'd unfortunately have to narrow both the argument and the function to make TS happy, like this:

```ts
if (Foo.is(item) && typeof payload == "string") {
  item.updatePayload(payload);
} else if (Bar.is(item) && Array.isArray(payload)) {
  item.updatePayload(payload);
} else {
  const type = getType(item);
  throw new Error(`Item type ${type.name} can't process payloads of type ${typeof payload}`);
}
```

Hope this helps! (and given this, I think we can close the issue?)

**Comment by coolsoftwaretyler** [comment]
Thanks @thegedge - that's very helpful. I'm going to convert this to a discussion, which will close the issue. Then I'll mark your answer as "the" answer.

Really appreciate your time, and your TS expertise.
